### PR TITLE
Support invoices without amount specified

### LINKED ISSUE #564: Support invoices without amount specified.
According to this:

https://github.com/lightningnetwork/lightning-rfc/blob/master/11-payment-encoding.md

amout is optional for payment invoice. Unfortunatelly lncli doesn't give option to create invoice without amout:

> lncli addinvoice
[lncli] value argument missing

In lncli there is sendpayment method that gives possibility for sender to specify amount, but I don't see command that creates payment_hash without invoice.

How can I query for payment without amout, and send payment  with amout specified by sender?

Also when trying to create invoice by RPC I got error message:

> Caused by: io.grpc.StatusRuntimeException: UNKNOWN: zero value invoices are disallowed

### PR DESCRIPTION
**Pull request by wilmerpaulino**
Fixes #564.

This PR introduces the following changes:

* Allows the creation of invoices without an amount.
* Handles incoming HTLCs paying an invoice without an amount.
* Allows payments using invoices without an amount, where the amount is specified at time of payment.
* Adds tests to decode and encode an invoice without an amount.

### CODE REVIEW

File: cmd/lncli/commands.go [hunk 1]
```diff
@@ -1133,9 +1140,10 @@ var addInvoiceCommand = cli.Command{
 	Usage: "add a new invoice.",
 	Description: `
 	Add a new invoice, expressing intent for a future payment.
-	
-	The value of the invoice in satoshis is necessary for the creation, 
-	the remaining parameters are optional.`,
+
+	Invoices without an amount can be created by not supplying any
+	parameters or providing an amount of 0. These invoices allow the payee
+	to specify the amount of satoshis they wish to send.`,
```
Comment (Roasbeef) [inline cmd/lncli/commands.go]: The `payinvoice` proxy command should be updated as well, since that's what most users default to these days.

**Comment by mably** [comment]
Looks like the payment value doesn't show up in the invoice settlement.  Or am I missing something?

**Comment by judezhu** [comment]
Yeah. Correct. The amount sent by the payment will NOT show up in the settled invoice.

**Comment by halseth** [comment]
The paid amount should show up as `AmtPaid`. If not please file an issue :)

**Pull request merged**